### PR TITLE
feat(sound): two-stage configurable sound detection for #12

### LINKED ISSUE #12: Task 8: Configurable Sound Detection Module
# Task 8: Configurable Sound Detection Module

- Block: **C — Audio Pipeline**
- Source of truth: this issue (inlined from `.sisyphus/plans/wooly-fluffy.md`)

## Dependencies
- #4 (Task 1: Project Scaffolding) — ✅ has PR
- #8 (Task 5: Shared Contracts)
- #34 (Task 8b: Audio HAL — provides AudioInput interface)

## Blocks
- #20 (Task 14: Activation System)
- #29 (Task 23: Debug Console + Sound Tuning UI)

---

## Specification

Implement a **2-stage configurable sound detector** that can detect arbitrary target sounds (bell as default, but user-tunable to any sound):

### Stage 1 — Cheap Gate
Bandpass energy filter on continuous audio stream. Only passes candidate segments with sufficient energy in target frequency range. Rejects silence and out-of-band noise cheaply.

### Stage 2 — Template Matcher
Log-mel spectrogram comparison against pre-recorded reference samples. Uses normalized cross-correlation (NCC) or cosine similarity on short candidate segments extracted by Stage 1.

### Extension (PANNs)
For harder-to-template sounds, provide alternative backend using PANNs embeddings + cosine similarity. Not required for MVP — activate when template matching proves insufficient.

### Calibration Workflow
- Staff records 5-10 reference samples of the target sound via Staff UI (Task 23) or CLI tool
- System extracts log-mel templates from samples, stores as `.npy` files in `data/sound_templates/`
- Calibration should include both clean samples and in-situ recordings (with ambient noise)
- Re-calibration is possible at any time without code changes

### Pattern Sound Detection (FSM)
- Support multi-event activation patterns (e.g., double clap, triple tap)
- FSM with configurable timing: `min_gap_ms`, `max_gap_ms`, `cooldown_ms`, `max_pattern_window_ms`
- Single-event confidence passed to FSM for multi-stage filtering
- Pattern definitions in settings.yaml under `sound_detection.activation_pattern`

### Audio Stream Processing
- Continuous microphone input via Audio HAL (Task 8b provides the interface)
- Ring buffer (16kHz, mono) with sliding window. Candidate-driven: Stage 1 triggers Stage 2 only on energy peaks
- Pre-computed templates loaded at startup; no per-frame STFT recomputation
- TTS echo suppression: `cooldown_after_tts_ms` window where detections are ignored (prevents self-triggering)

### Output
`SoundDetected(confidence: float, source: str, sound_type: str, timestamp: datetime)` event

### Configurable Parameters (via settings.yaml)
`sound_detection.sensitivity_threshold`, `sound_detection.min_target_hz`, `sound_detection.max_target_hz`, `sound_detection.debounce_ms`, `sound_detection.backend`, `sound_detection.template_dir`, `sound_detection.activation_pattern.max_pattern_window_ms`

### Performance Target
Detection within 500ms of sound onset.

### API Endpoints (registered on FastAPI app from Task 6)

- `GET /api/sound/status` — Current detector status. Response: `{"active": true, "last_detection": "<ISO|null>", "confidence_threshold": <float>, "detections_today": <int>, "backend": "template|panns", "templates_loaded": <int>}`
- `PUT /api/sound/sensitivity` — Update sensitivity at runtime. Body: `{"threshold": <float 0.0-1.0>}`. Response: `{"threshold": <new_value>, "applied": true}`
- `GET /api/sound/history` — Recent detection events. Query params: `?limit=20`. Response: `[{"timestamp": "<ISO>", "confidence": <float>, "source": "gate+template", "sound_type": "bell"}]`
- `POST /api/sound/test` — Test sound detection pipeline. Two modes: (a) No body → publish a synthetic SoundDetected event to event bus (for integration testing). (b) Multipart audio file body → feed audio through the actual detection pipeline (gate + template matching) and return real detection result. Response: `{"event_id": "<id>", "mode": "synthetic|pipeline", "detected": true|false, "confidence": <float|null>, "message": "..."}`
- `POST /api/sound/calibrate` — Accept uploaded audio samples for template extraction. Body: multipart audio file. Response: `{"template_id": "<id>", "features_extracted": true, "total_templates": <int>}`

### Must NOT Do
- Do not implement camera-based activation (Task 12 + Task 14)
- Do not implement IoT button integration (future enhancement)
- Do not use YAMNet or TensorFlow as runtime dependency (PANNs via PyTorch is the ML extension path)
- Do not implement real-time software beamforming (use audio from HAL as-is)
- Do not store raw audio recordings beyond processing buffer (privacy constraint)

---

## Files to Create

| File | Description |
|------|-------------|
| `daemon/src/daemon/sound/__init__.py` | Module init with `__all__` exports |
| `daemon/src/daemon/sound/detector.py` | 2-stage detection pipeline |
| `daemon/src/daemon/sound/gate.py` | Stage 1 bandpass energy gate |
| `daemon/src/daemon/sound/template_matcher.py` | Stage 2 log-mel spectrogram template matching |
| `daemon/src/daemon/sound/panns_backend.py` | PANNs embedding backend (extension) |
| `daemon/src/daemon/sound/calibration.py` | Template extraction and calibration logic |
| `daemon/src/daemon/sound/pattern_fsm.py` | Multi-event pattern FSM |
| `daemon/src/daemon/sound/routes.py` | FastAPI route definitions for sound endpoints |
| `tests/sound/__init__.py` | Test package init |
| `tests/sound/test_gate.py` | Stage 1 gate tests |
| `tests/sound/test_template_matcher.py` | Stage 2 template matching tests |
| `tests/sound/test_detector.py` | Full pipeline tests |
| `tests/sound/test_pattern_fsm.py` | Pattern FSM timing tests |
| `tests/sound/test_calibration.py` | Calibration API tests |
| `tests/sound/test_routes.py` | API endpoint tests |
| `fixtures/audio/bell_clean.wav` | Clean bell audio fixture |
| `fixtures/audio/not_bell_metallic.wav` | Non-bell metallic sound fixture |
| `fixtures/audio/speech_only.wav` | Speech-only audio fixture |
| `data/sound_templates/.gitkeep` | Template storage directory |

## Existing Codebase Conventions
- Import pattern: `from daemon.sound import ...`
- Module init: `__all__` export pattern (see `daemon/src/daemon/domain/__init__.py`)
- Python 3.14, `from __future__ import annotations` in every file
- `asyncio_mode = "auto"` in pytest config
- Per-file-ignores may be needed in pyproject.toml

---

## Acceptance Criteria

- [ ] TDD: `pytest tests/sound/` passes with ≥10 tests
- [ ] Stage 1 gate: passes bell audio, rejects silence and speech-only audio
- [ ] Stage 2 template match: detects target sound with ≥90% confidence using 5+ calibration samples
- [ ] Detects bell-with-noise fixture with ≥75% confidence
- [ ] Rejects non-target metallic sounds (false positive rate <10%)
- [ ] Detection latency ≤500ms from sound onset
- [ ] Configurable sensitivity via settings.yaml
- [ ] Calibration API: accepts audio upload, extracts template, stores .npy
- [ ] Pattern FSM: double-event pattern detected with correct timing constraints
- [ ] TTS echo suppression: detection suppressed during cooldown_after_tts_ms window

## QA Scenarios

    Scenario: Two-stage bell detection on clean audio fixture
      Tool: Bash (pytest)
      Preconditions: Calibration templates loaded from fixtures/audio/bell_samples/, detector initialized
      Steps:
        1. Load fixtures/audio/bell_clean.wav
        2. Feed audio through Stage 1 gate — assert candidate segment extracted (energy above threshold in 2-8kHz)
        3. Feed candidate through Stage 2 template matcher — assert NCC score ≥ 0.90 against reference templates
        4. Assert SoundDetected event emitted with confidence ≥ 0.90, sound_type="bell"
        5. Assert detection timestamp within 500ms of audio onset
      Expected Result: High-confidence detection through both stages within latency target
      Failure Indicators: Gate rejects bell audio, template match too low, high latency
      Evidence: .sisyphus/evidence/task-8-bell-clean.txt

    Scenario: False positive rejection on non-target sounds
      Tool: Bash (pytest)
      Preconditions: Bell templates loaded, non-bell audio fixtures exist (dropped keys, clapping, toy sounds)
      Steps:
        1. Load fixtures/audio/not_bell_metallic.wav (metallic but not bell)
        2. Feed through detector — if Stage 1 passes, Stage 2 should reject (NCC < threshold)
        3. Load fixtures/audio/speech_only.wav
        4. Assert Stage 1 gate rejects (insufficient energy in target band)
        5. Assert NO SoundDetected events emitted for either fixture
      Expected Result: Non-target sounds correctly rejected at either stage
      Failure Indicators: False positive SoundDetected event
      Evidence: .sisyphus/evidence/task-8-false-positive.txt

    Scenario: TTS echo suppression prevents self-triggering
      Tool: Bash (pytest)
      Preconditions: Detector running, cooldown_after_tts_ms=2000
      Steps:
        1. Emit TTSSpeechStarted event to event bus
        2. Immediately feed bell audio to detector
        3. Assert NO SoundDetected event emitted (within cooldown window)
        4. Wait 2100ms after TTSSpeechFinished
        5. Feed same bell audio again
        6. Assert SoundDetected event IS emitted
      Expected Result: Detections suppressed during TTS playback + cooldown, resume after
      Failure Indicators: Detection during cooldown, no detection after cooldown
      Evidence: .sisyphus/evidence/task-8-echo-suppression.txt

    Scenario: Pattern sound detection (double tap)
      Tool: Bash (pytest)
      Preconditions: settings.yaml sound_detection.activation_pattern.type="double", min_gap_ms=200, max_gap_ms=800
      Steps:
        1. Feed single tap audio → assert NO activation event (single event stored in FSM)
        2. Feed second tap audio 400ms later → assert SoundDetected event with pattern="double"
        3. Feed two taps 50ms apart (too fast) → assert NO activation (gap < min_gap_ms)
        4. Feed two taps 1200ms apart (too slow) → assert NO activation (gap > max_gap_ms)
      Expected Result: Pattern FSM correctly validates timing constraints
      Failure Indicators: Activation on single event, wrong timing acceptance/rejection
      Evidence: .sisyphus/evidence/task-8-pattern-fsm.txt

    Scenario: Calibration API template extraction
      Tool: Bash (curl)
      Preconditions: Daemon running, /api/sound/calibrate endpoint available
      Steps:
        1. POST /api/sound/calibrate with bell_sample_1.wav — assert 200, features_extracted=true
        2. POST /api/sound/calibrate with bell_sample_2.wav — assert total_templates=2
        3. GET /api/sound/status — assert templates_loaded=2
        4. Verify .npy files created in data/sound_templates/
      Expected Result: Audio samples converted to templates and persisted
      Failure Indicators: Extraction failure, no .npy files, wrong template count
      Evidence: .sisyphus/evidence/task-8-calibration.txt

---

## TDD Workflow
1. **RED**: Create test files with imports that don't exist yet
2. **GREEN**: Implement minimal code to pass
3. **REFACTOR**: Improve while keeping tests green
4. **Verify**: Run quality gates

## Quality Gates

    uv run ruff format --check daemon/src/daemon/sound tests/sound
    uv run ruff check daemon/src/daemon/sound tests/sound
    uv run mypy --strict daemon/src tests
    uv run pyright daemon/src
    uv run pytest tests/sound/ -v --tb=short

## ADR Requirement
- `adr_required: false`
- Rationale: ADR 0004 already covers template matching over YAMNet decision
- No new ADR-triggering decisions needed

## Commit Message

    feat(sound): two-stage configurable sound detection with template matching, pattern FSM, and calibration API

## External References
- librosa log-mel spectrogram: https://librosa.org/doc/latest/generated/librosa.feature.melspectrogram.html — Template feature extraction reference (use for offline calibration; runtime uses pre-computed NumPy arrays)
- PANNs inference: https://github.com/qiuqiangkong/panns_inference — Extension path for embedding-based sound matching. CNN14 model recommended for balance of accuracy and speed.
- sounddevice library: https://python-sounddevice.readthedocs.io/ — Continuous audio stream on macOS (used by Audio HAL, but detector should understand stream format)
- NumPy cross-correlation: `np.correlate` / `scipy.signal.correlate` — For NCC-based template matching at runtime

### PR DESCRIPTION
**Pull request by ToaruPen**
Closes #12

## Summary by CodeRabbit

* **新機能**
  * サウンド検出設定を拡張：周波数範囲、デバウンス、ゲート閾値、テンプレート格納先、検出バックエンド選択、感度既定値、発火パターン（単/複数）とタイミング制御を追加。
  * ランタイムでサウンド検出サービスを起動・停止し、HTTP APIで状態、感度更新、校正、テスト、履歴を操作可能に。
  * テンプレート校正、マッチング、埋め込みバックエンド、パターンFSM、バンドパスゲートを導入。

* **テスト**
  * サウンド関連の幅広い単体・統合テストを追加。

* **ドキュメント**
  * ADRにサウンド検出の決定記録を追記。

### CODE REVIEW

**Comment by ToaruPen** [comment]
@coderabbitai resolve

**Comment by ToaruPen** [comment]
@coderabbitai review

**Comment by ToaruPen** [comment]
@coderabbitai resolve
